src/nodes.coffee

```diff
@@ -268,41 +272,101 @@ exports.Base = class Base
   # Plain JavaScript object representation of the node, that can be serialized
   # as JSON. This is used for generating an abstract syntax tree (AST).
   # This is what the `ast` option in the Node API returns.
-  toJSON: ->
+  toJSON: (o) ->
     # We try to follow the [Babel AST spec](https://github.com/babel/babel/blob/master/packages/babylon/ast/spec.md)
     # as closely as possible, for improved interoperability with other tools.
-    obj =
-      type: @constructor.name
-      # Convert `locationData` to Babel’s style.
-      loc:
-        start:
-          line: @locationData.first_line
-          column: @locationData.first_column
-        end:
-          line: @locationData.last_line
-          column: @locationData.last_column
-
-    # Add serializable properties to the output. Properties that aren’t
-    # automatically serializable (because they’re already a primitive type)
-    # should be handled on a case-by-case basis in child node classes’ own
-    # `toJSON` methods.
-    for property, value of this
-      continue if property in ['locationData', 'children']
-      continue if value is undefined # Don’t skip `null` or `false` values.
-      if typeof value is 'boolean' or typeof value is 'number' or typeof value is 'string'
-        obj[property] = value
-
-    # Work our way down the tree. This is like `eachChild`, except that we
-    # preserve the child node name, and arrays.
-    for attr in @children when @[attr]
-      if Array.isArray(@[attr])
-        obj[attr] = []
-        for child in flatten [@[attr]]
-          obj[attr].push child.unwrap().toJSON()
-      else
-        obj[attr] = @[attr].unwrap().toJSON()
+    @toAst o
```

Comment:
If `toJSON` is meant as an external API, how can it take an argument `o`?

What’s the point of having a method that just calls another method? Why not put the body of `toAst` in here?

Reply:
None of the `nodes.coffee` classes or their methods are considered “public.” The only public-facing API is the `CoffeeScript` object documented in https://coffeescript.org/#nodejs-usage.

I was thinking of `toJSON` as the AST equivalent of `compileNode`. `compileNode` is called on a node to get its JavaScript string representation, and `compileNode` itself often calls other methods on the node (and child nodes’ `compileNode` methods) as part of assembling that string output. When I was initially starting to sketch out what an AST generation would look like, I was imagining a `compileASTNode` to make the equivalence explicit, which later got renamed to `toJSON`.

I understand the appeal of keeping the AST generation logic encapsulated into its own dedicated methods on the classes, but I think it’s more important to follow the general pattern of how these classes are set up, and how the methods override each other. I think the idea should be that when the time comes for `nodes.coffee` to only output AST, and then a separate printer file turns that AST into JavaScript, what we would do is remove the nodes’ `compileNode` and related (`compileToFragments` etc.) methods, but the rest would remain.

Reply:
> None of the `nodes.coffee` classes or their methods are considered “public.” The only public-facing API is the `CoffeeScript` object documented in https://coffeescript.org/#nodejs-usage.

@GeoffreyBooth I know, I'm trying to be clear that I only mean "external"/"public" in the sense of "external to `nodes.coffee`", not external to Coffeescript. But then given that, why should we even use the name `toJSON()` if it's only an internally-used name? My point is basically that the names should be helpful to someone working inside our codebase. I think that having AST-related functionality named "AST" is much easier to follow that having to say "oh ok, `toJSON()` is what you call to get an AST". My understanding is that the idea to use the name `toJSON()` came from it being a "standard" that `JSON.stringify()` respects. But so that's a name that makes sense to someone *consuming* our node class objects (ie someone "external"), it doesn't make as much sense "internally" (eg when working with the `nodes.coffee` codebase)

> I was thinking of `toJSON` as the AST equivalent of `compileNode`

So whether it's called `toJSON()` or `toAst()`, it is quite similar to how `compileNode()` works. In a general sense, there's a default implementation on the base class and then different node classes can override a couple different methods to customize the way that it takes responsibility for generating JS or AST for itself and its child nodes recursively

I'd make a rough comparison:
- `compileToFragments()` <---> `toAst()` (outermost, not frequently overridden as it mostly contains common logic and then relies on the next level down for the core node-type-specific rendering logic)
- `compileNode()` <---> `_toAst()` (responsible for the node-type-specific rendering logic)
- and then `children` <---> `astChildren`/`astProps`/`astType` (they're at least conceptually similar, I don't think `children` gets used as part of default JS generation but imagine if the default `compileNode()` eg concatenated the JS strings of the `children`)

> I understand the appeal of keeping the AST generation logic encapsulated into its own dedicated methods on the classes, but I think it’s more important to follow the general pattern of how these classes are set up, and how the methods override each other

I just described how I think we're following existing patterns. I'm not sure exactly what you're picturing as far as "keeping the AST generation logic encapsulated into its own dedicated methods on the classes" but I think it will be similar to how `compileNode()` etc work in that the specific implementations of those known methods on the node classes will often call out to other helper methods (whether defined on the node class or elsewhere). In fact, part of what I've started doing now that I know we're targeting a version where existing `compileNode()` logic and AST generation logic will sit side-by-side is to extract more logic that needs to be shared between the two into helper methods